### PR TITLE
Tuning persistent batchnorm

### PR DESCRIPTION
**Pull request by naoyam**
This PR is not for proposing changes but for demonstrating how persistent batchnorm can be tuned.

`BWTest3` is the batchnorm test tuned for Titan RTX. Here's the result on my system:

`PYTORCH_NVFUSER_DUMP=cuda_to_file,ptxas_verbose,fusion_ir_math,dump_eff_bandwidth    ./build/bin/test_jit  --gtest_filter="NVFuserTest.*BWTest3"`

```
ptxas info    : 57 bytes gmem
ptxas info    : Compiling entry function '_ZN11CudaCodeGen7kernel1ENS_6TensorINS_6__halfELi4EEENS0_IS1_Li1EEES3_NS0_IfLi1EEES4_S4_S4_S2_S4_S4_NS0_IiLi1EEENS0_IxLi1EEE' for 'sm_75'
ptxas info    : Function properties for _ZN11CudaCodeGen7kernel1ENS_6TensorINS_6__halfELi4EEENS0_IS1_Li1EEES3_NS0_IfLi1EEES4_S4_S4_S2_S4_S4_NS0_IiLi1EEENS0_IxLi1EEE
ptxas         .     0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 123 registers, 16 bytes smem, 592 bytes cmem[0], 8 bytes cmem[2]

kernel1 run in 0.50713598728179932 ms, achieved: 405.2983443389262 GB/s
kernel1 run in 0.49568000435829163 ms, achieved: 414.66545794215426 GB/s
kernel1 run in 0.51631999015808105 ms, achieved: 398.08913061272267 GB/s
kernel1 run in 0.51078397035598755 ms, achieved: 402.40373216244291 GB/s
kernel1 run in 0.4981440007686615 ms, achieved: 412.61437592912739 GB/s
kernel1 run in 0.53763198852539062 ms, achieved: 382.30868026241512 GB/s
kernel1 run in 0.50774401426315308 ms, achieved: 404.81299675838659 GB/s
kernel1 run in 0.51964801549911499 ms, achieved: 395.53961502687599 GB/s
kernel1 run in 0.50352001190185547 ms, achieved: 408.20895126619808 GB/s
kernel1 run in 0.52012801170349121 ms, achieved: 395.17459428271047 GB/s
kernel1 run in 0.50003200769424438 ms, achieved: 411.05643806242659 GB/s
```

Originally, it was as low as 100 GB/s. 

Some hacky changes are required for now:

1. Manual setting of computeAt positions of welford output tensors. `setComputeAt` and `setMaxProducer` are changed to be public functions. There's an alternative (hacky) way that doesn't require the functions to be public.
2. Bypassing vectorization check. Two inner-most dimensions, `[14, 14]`, are merged and then vectorized, which should allow 4-way vectorization, but the current vectorization validation looks at each root axis separately, so only 2-way is allowed.

Another changer that is not required but apparently beneficial is to do exponential backoff of conflicting grid sync. Just copied the code from https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#scheduling-example.

Also, commented out some runtime files for the complex type. It seems they interfere when generated code is given to nvcc.

Other things I tried but not effective:

- Warp reduction. This particular problem does block reductions with the thread block shape of `[49, 8]` and  only the TIDx dimension is reduced. I manually experimented with warp reductions. Got about 5 to 10 % improvement with the whole grid fused welford, but almost negligible in the overall batchnorm.
- Double buffering. Tried double buffering of the input. Resulted in failure of register allocation due to dynamic indexing caused by the use of the index of the outer-most loop, which is not unrolled. Hacked codegen to workaround and got all of them allocated on registers, but still the resulting perf was significantly slower.

Some other things I considered but not tried:
- Prefetching. PTX has prefetch. Helpful? I've never used it, though.
- Cache the input on shared memory on A100, which has much larger shared memory than Turing.

### CODE REVIEW

**Comment by csarofeen** [comment]
@naoyam I think we can probably close this.
